tools/filterdiff.py

```diff
@@ -0,0 +1,133 @@
+# /usr/bin/env python3
+import sys
+import os
+import argparse
+import tempfile
+import json
+import subprocess
+from typing import List, Dict, Optional, Set, Tuple
+
+
+class FilterImageInfo:
+    """
+    Filter specific fields in the image-info that can't be compared
+    together.
+    """
+
+    def __init__(self, imi: Optional[Dict], to_remove_tmpfiles_d: Set) -> None:
+        self.imi: Dict = imi if imi else {}
+        self.to_remove_tmpfiles_d = to_remove_tmpfiles_d
+
+    def _tmpfiles_d(self):
+        """
+        Remove certain files from the tpmfiles_d that have a tendency of having
+        a non fixed content.
+        """
+        tmpfilesd = self.imi.get("tmpfiles.d")
+        if tmpfilesd:
+            for _, v in tmpfilesd.items():
+                for trm in self.to_remove_tmpfiles_d:
+                    if trm in v:
+                        del v[trm]
+
+    def _lvm2(self):
+        """
+        LVM2 partitions have a UUID that is not fixed. Replace the value
+        upon comparison time.
+        """
+        partitions = self.imi.get("partitions")
+        if partitions:
+            for partition in partitions:
+                if partition.get("fstype") == "LVM2_member":
+                    partition["uuid"] = "2022-07-01-fixed-uuid"
+
+    def _iso(self):
+        """
+        For isos, the partition UUID is the date of the build. Replace
+        that with a fixed one for the comparison.
+        """
+        if ("image-format" in self.imi and "type" in self.imi["image-format"] and
+                self.imi["image-format"]["type"] == "raw"):
+            if "partitions" in self.imi:
+                for partition in self.imi["partitions"]:
+                    if "fstype" in partition:
+                        if partition["fstype"] == "iso9660":
+                            partition["uuid"] = "2022-07-01-fixed-uuid"
+
+    def apply(self) -> Dict:
+        self._iso()
+        self._lvm2()
+        self._tmpfiles_d()
+        return self.imi
+
+
+def load(path) -> Tuple[Dict, Set]:
+    """
+    Returns the image-info bits from the input path. If it exit an
+    unwanted_tmpfiles_d in the file, return it also=.
+    """
+    c = None
+    with open(path, "r") as f:
+        c = json.load(f)
+    # Supports raw image-info or DB entry
+    if "image-info" in c:
+        c = c["image-info"]
+    unwanted_tmpfiles_d = set(c.get("unwanted_tmpfiles_d", []))
+    return c, unwanted_tmpfiles_d
```

Comment:
Thank you, I very appreciate your rework.

At the same time I'm sorry, but I don't like the end result. My reasons are:

- You may end up having to filter more than just tmpfilesd entries in the future and in such case the `load_item()` number of return values will just explode.
- The line `c = c["image-info"]` will always override what is in the `c` with `image-info` key value if it exists, therefore there won't be any `unwanted_tmpfiles_d` key if this is a DB entry.

It seems to me that many the approach of trying to handle multiple situations at once unnecessarily complicates the logic.

Both `diff()` and `FilterImageInfo()` seem to be operating only on `image-info` report if it exists in the loaded file.

I would propose to simply return the `dict` as loaded from the file. Then in the `diff()` get the `unwanted_tmpfiles_d` from each entry + something like `imi = c.get("image-info", [])`. Run the returned list through `FilterImageInfo()` and then diff the result. If you feel bad about duplicating 3 lines, you can wrap this pre-processing in the `load()` or a different function and return the end result for diffing.

Reply:
There is great value at handling the two situations. The reason why:
* You got an comparison mismatch on the CI, you wan't to check what's going on, you download the image info and run the comparison with what's int he DB. You then need to be able to compare a DB entry with an image-info.
So I insist on the need to handle this two usecases.

Good catch for the unwanted override. Too much rewrites lead to meh code sometimes.

Reply:
I agree and do not dispute the usefulness of handling the two situations by the code. What I'm trying to suggest is to concentrate the differences ideally in a single function and then if possible, make the rest of the code agnostic to the fact if the input was image-info report of a DB entry. Which you kind of already have. My suggestion is to basically have all the double situation handling in `diff()`. I admit that there is no ideal solution, so I'll be happy with any outcome that does not have the bug from rewrite...

Reply:
Thanks for clarifying :smile: I'm gonna give this a bit more thoughts then.